src/main.rs

```diff
@@ -122,15 +195,33 @@ fn process_solve(theory: &Theory, bound: Option<BoundCommand>, count: Option<i32
         None
     };
 
-    strategy.add(Model::new(), selector);
     let mut found = 0;
 
-    while !strategy.empty() {
-        if count.is_some() && found >= count.unwrap() {
-            break;
+    // FIXME: use dynamic dispatching
+    match strategy {
+        StrategyOption::FIFO => {
+            let mut strategy = FIFO::new();
+            strategy.add(Model::new(), selector);
+
+            while !strategy.empty() {
+                if count.is_some() && found >= count.unwrap() {
+                    break;
+                }
+                solve(&mut strategy, &evaluator, bounder.as_ref(), &mut |m| { print_model(m, color, &mut found) })
```

Comment:
Seeing a mutable reference to a closure is rather weird, what's the type signature of this `solve` function?

Reply:
true- I'm not sure why I did that but I'll fix it.

Reply:
`pub fn solve<'s, S, M, Sel, Stg, E, B>(strategy: &mut Stg, evaluator: &E, bounder: Option<&B>, consumer: &mut impl FnMut(M))`

Reply:
```
pub fn solve<'s, S, M, Sel, Stg, E, B>(strategy: &mut Stg, evaluator: &E, bounder: Option<&B>, consumer: &mut impl FnMut(M))
    where S: 's + SequentTrait,
          M: ModelTrait,
          Sel: SelectorTrait<Item=&'s S>,
          Stg: StrategyTrait<'s, S, M, Sel>,
          E: EvaluatorTrait<'s, Sel, B, Sequent=S, Model=M>,
          B: BounderTrait
```

Reply:
the compiler doesn't let me to change the signature to an immutable reference. I guess it has something to do with the closure being `FnMut`?

Reply:
Can you change that `consumer` parameter to be `mut consumer: impl FnMut(M)` instead?

Reply:
you're right. that works! I remember I needed a reference to the closure so I could pass it to the function multiple times but seems like I was missing something.
Thanks for the suggestion!

Reply:
Ok, now that's solved, you should then be able to remove the `&mut` on this closure and it should still compile.